internal/provider/resource_access_model.go

```diff
@@ -452,38 +458,106 @@ func (o *accessModelResource) UpgradeState(ctx context.Context) map[int64]resour
+				stateDataV2 := o.stateDataV1ToV2(ctx, resp, stateDataV1)
+				if resp.Diagnostics.HasError() {
 					return
 				}
 
-				upgradedStateData := accessModelResourceModelV1{
-					accessModelResourceModel: accessModelResourceModel{
-						Model:  priorStateData.Model,
-						Access: priorStateData.Access,
-						ID:     priorStateData.ID,
+				resp.Diagnostics.Append(resp.State.Set(ctx, stateDataV2)...)
+			},
+		},
+		// Upgrade from model to model_uuid
+		1: {
+			PriorSchema: &schema.Schema{
+				Attributes: map[string]schema.Attribute{
+					"model": schema.StringAttribute{
+						Required: true,
 					},
-					Users: usersSet,
+					"users": schema.SetAttribute{
+						Required:    true,
+						ElementType: types.StringType,
+					},
+					"access": schema.StringAttribute{
+						Required: true,
+					},
+					"id": schema.StringAttribute{
+						Computed: true,
+					},
+				},
+			},
+			StateUpgrader: func(ctx context.Context, req resource.UpgradeStateRequest, resp *resource.UpgradeStateResponse) {
+				var stateDataV1 accessModelResourceModelV1
+
+				resp.Diagnostics.Append(req.State.Get(ctx, &stateDataV1)...)
+				if resp.Diagnostics.HasError() {
+					return
 				}
 
-				resp.Diagnostics.Append(resp.State.Set(ctx, upgradedStateData)...)
+				stateDataV2 := o.stateDataV1ToV2(ctx, resp, stateDataV1)
+
+				resp.Diagnostics.Append(resp.State.Set(ctx, stateDataV2)...)
 			},
 		},
 	}
 }
+
+func (o *accessModelResource) stateDataV0ToV1(ctx context.Context, resp *resource.UpgradeStateResponse, priorStateData accessModelResourceModelV0) accessModelResourceModelV1 {
+	users := []string{}
+	if !priorStateData.Users.IsNull() {
+		resp.Diagnostics.Append(priorStateData.Users.ElementsAs(ctx, &users, false)...)
+	}
+
+	usersSet, diags := types.SetValueFrom(ctx, types.StringType, users)
+	if diags.HasError() {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to convert users to set, got error: %s", diags))
+		return accessModelResourceModelV1{}
+	}
+
+	return accessModelResourceModelV1{
+		accessModelResourceModel: accessModelResourceModel{
+			Access: priorStateData.Access,
+			ID:     priorStateData.ID,
+		},
+		Model: priorStateData.Model,
+		Users: usersSet,
+	}
+}
+
+func (o *accessModelResource) stateDataV1ToV2(ctx context.Context, resp *resource.UpgradeStateResponse, priorStateData accessModelResourceModelV1) accessModelResourceModelV2 {
+	modelStr := priorStateData.Model.ValueString()
+	modelUUID, err := o.client.Models.ModelUUID(modelStr)
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to get model UUID for model %q, got error: %s", modelStr, err))
+		return accessModelResourceModelV2{}
+	}
+
+	users := []string{}
+	if !priorStateData.Users.IsNull() {
+		resp.Diagnostics.Append(priorStateData.Users.ElementsAs(ctx, &users, false)...)
+	}
+
+	newID := newAccessModelIDFrom(modelUUID, priorStateData.Access.ValueString(), users)
```

Comment:
Now this might be a problem right (and maybe elsewhere too), because of what we were discussing that users is a set, so when you unmarshal to a list you may have a different order.
We can opt to do a strings.Replace() on the old ID (see some of the other upgrade state methods) to be safe.

Reply:
Ah that's good then. It feels a bit shaky to rely on that. If they changed it to a map and we get back a random order it would suck but like you mentioned, that would probably break many providers.